### PR TITLE
Call reload_model_params only if there's no optimizer state

### PR DESCRIPTION
**Pull request by akoumpa**
# What does this PR do ?

Checkpoint restoration process (roughly):
1. generate a state dict that contains model's state dict and optimizer's state dict.
2. load the two state dicts from the file checkpoint
3. restore model's state from the state contained in the checkpoint
4. reload optimizer params from the model's state.
5. restore optimizer's state from the state contained in the checkpoint.

The bug was in (1) and (4), the optimizer state does not contain cloned tensors but rather references. As a result, when the parameters were reloaded in (4) they were actually overwriting what was restored in (2) (in particular the fp32_from_fp16_params).

**Collection**: [Note which collection this PR will affect]

# Changelog 
- Add specific line by line info of high level changes in this PR.

# Usage
* You can potentially add a usage example below

```python
# Add a code snippet demonstrating how to use this 
```

# GitHub Actions CI

The Jenkins CI system has been replaced by GitHub Actions self-hosted runners.

The GitHub Actions CI will run automatically when the "Run CICD" label is added to the PR.
To re-run CI remove and add the label again.
To run CI on an untrusted fork, a NeMo user with write access must first click "Approve and run".

# Before your PR is "Ready for review"
**Pre checks**:
- [ ] Make sure you read and followed [Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md)
- [ ] Did you write any new necessary tests?
- [ ] Did you add or update any necessary documentation?
- [ ] Does the PR affect components that are optional to install? (Ex: Numba, Pynini, Apex etc)
  - [ ] Reviewer: Does the PR have correct import guards for all optional libraries?
  
**PR Type**:
- [ ] New Feature
- [ ] Bugfix
- [ ] Documentation

If you haven't finished some of the above items you can still open "Draft" PR.

## Who can review?

Anyone in the community is free to review the PR once the checks have passed. 
[Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md) contains specific people who can review PRs to various areas.

# Additional Information
* Related to # (issue)

### CODE REVIEW

File: nemo/lightning/pytorch/strategies/megatron_strategy.py [hunk 1]
```diff
@@ -715,8 +715,10 @@ def load_model_state_dict(self, checkpoint: Mapping[str, Any], strict: bool = Tr
         assert self.megatron_parallel is not None
 
         _strategy_lib.load_model_state_dict(self.megatron_parallel, checkpoint, strict=strict)
-        for opt in self.optimizers:
-            opt.reload_model_params()
+
+        if not 'optimizer' in checkpoint:
```
Comment (hemildesai) [inline nemo/lightning/pytorch/strategies/megatron_strategy.py:719]: Should this also check for `optimizer_states` key?

Comment (akoumpa) [reply nemo/lightning/pytorch/strategies/megatron_strategy.py:719]: I assumed it would be empty https://github.com/NVIDIA/NeMo/blob/main/nemo/lightning/pytorch/strategies/megatron_strategy.py#L644

Comment (hemildesai) [reply nemo/lightning/pytorch/strategies/megatron_strategy.py:719]: Makes sense 👍

**Review by hemildesai — DISMISSED** [review]

**Review by hemildesai — APPROVED** [review]
